### ISSUE TITLE
Align with fractal-server 1.4.0

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-web#335]
**[this issue is still in progress]**

Main changes in fractal-server 1.4:

1. There is an important change which does not require immediate action in fractal-web but that we'll need to keep in mind. It concerns the ApplyWorkflow (AKA "Job") table, where the four existing foreign keys (input_dataset_id, output_dataset_id, workflow_id, project_id) become optional. This means that from now on there will be different types of jobs:
    * A job with all FKs non-null, which is in relation with a workflow, a project, and one or two datasets
    * A job which is still in relation with a project (and therefore accessible to the user), but possibly not with a workflow (or with a dataset). This is the typical situation when a workflow (or a dataset) is removed after the job already exists.
    * A job which has all four FKs set to null, which happens when the user deletes a whole project. This kind of jobs are meant to be an archive, which is only accessible to a superuser for accounting reasons.
2. In some cases, when looking at jobs which have some null FK it will be useful to learn something about the workflow or dataset(s) that they were related to at creation time. This will be possible via the following job properties: `workflow_dump`, `input_dataset_dump`, `output_dataset_dump`, that are (obviously) dumps of the corresponding DB items taken at the job-creation time.
3. In `GET /api/v1/project/` and `GET /api/v1/project/{project_id}`, project acquires new properties `workflow_list` and `job_list` (while `dataset_list` was already present in previous versions).
4. There will be a new set of superuser-restricted monitoring endpoints (https://github.com/fractal-analytics-platform/fractal-server/pull/947), to be used as part of https://github.com/fractal-analytics-platform/fractal-web/issues/301.
5. There will be a couple of new endpoints for getting jobs at the per-workflow and per-user level (https://github.com/fractal-analytics-platform/fractal-server/issues/920)

I'll continue later, and include more specific details of things that may require changes in fractal-web.

**Comment by tcompa** [comment]
Here is a more structured CHANGELOG, corresponding to the first prerelease (1.4.0a0). Note that a few more changes are expected, before 1.4.0.

---

* API:
    * Make it possible to delete a `Dataset`, `Workflow` or `Project`, even when it is in relationship to an `ApplyWorkflow` (\#927).
    * Include `workflow_list` and `job_list` in `ProjectRead` response (\#927).
* Database:
    * Make foreign-keys of `ApplyWorkflow` (`project_id`, `workflow_id`, `input_dataset_id`, `output_dataset_id`) optional (\#927).
    * Add columns `input_dataset_dump`, `output_dataset_dump` and `user_email` to `ApplyWorkflow` (\#927).
    * Add relations `Dataset.list_jobs_input` and `Dataset.list_jobs_output` (\#927).
    * Make `ApplyWorkflow.start_timestamp` non-nullable (\#927).
    * Remove `"cascade": "all, delete-orphan"` from `Project.job_list` (\#927).
    * Add `Workflow.job_list` relation (\#927).
    *  (\#927).
* Runner:
    * Refresh DB objects within `submit_workflow` (\#927).

**Comment by tcompa** [comment]
Version `1.4.0a1` is now available.

Here is the up-to-date changelog:

* API:
    * New endpoints:
        * New monitoring endpoints restricted to superusers at `/monitoring` (\#947).
        * New `GET` endpoints `api/v1/project/job/` and `api/v1/project/{project_id}/workflow/{workflow_id}/job/` (\#969).
    * New behaviors or responses of existing endpoints:
        * Change response of `/api/v1/project/{project_id}/job/{job_id}/stop/` endpoint to 204 no-content (\#967).
        * Include `workflow_list` and `job_list` attributes for `ProjectRead`, which affects all `GET`-project endpoints (\#927).
        * Make it possible to delete a `Dataset`, `Workflow` or `Project`, even when it is in relationship to an `ApplyWorkflow` (\#927).
    * Internal changes:
        * Fix construction of `ApplyWorkflow.workflow_dump`, within apply endpoint (\#968).
* Database:
    * Make foreign-keys of `ApplyWorkflow` (`project_id`, `workflow_id`, `input_dataset_id`, `output_dataset_id`) optional (\#927).
    * Add columns `input_dataset_dump`, `output_dataset_dump` and `user_email` to `ApplyWorkflow` (\#927).
    * Add relations `Dataset.list_jobs_input` and `Dataset.list_jobs_output` (\#927).
    * Make `ApplyWorkflow.start_timestamp` non-nullable (\#927).
    * Remove `"cascade": "all, delete-orphan"` from `Project.job_list` (\#927).
    * Add `Workflow.job_list` relation (\#927).
    * Do not use `Enum`s as column types (e.g. for `ApplyWorkflow.status`), but only for (de-)serialization (\#974).
* Runner:
    * Refresh DB objects within `submit_workflow` (\#927).
* Testing:
    * Improve `test_full_workflow.py` (\#971).

**Comment by tcompa** [comment]
This is the list of issues that in principle should be part of 1.4.0 (or of an upcoming patch version):
https://github.com/fractal-analytics-platform/fractal-server/issues?q=is%3Aissue+is%3Aopen+label%3A1.4.0

**Comment by tcompa** [comment]
The fractal-server prereleases 1.4.0a8 is now available, and most likely think nothing else will change in the API between this one and 1.4.0.

Here is a recap of recent updates, from a6 to a7 and from a7 to a8.

## `git diff 1.4.0a6...1.4.0a7 CHANGELOG.md`

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 38a8f1cc..6f31b5be 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,7 +5,7 @@
 * API:
     * Major endpoint changes:
         * Add trailing slash to _all_ endpoints' paths (\#1003).
-        * Add new monitoring endpoints restricted to superusers at `/monitoring` (\#947).
+        * Add new admin-area endpoints restricted to superusers at `/admin` (\#947, \#1009, \#1032).
         * Add new `GET` endpoints `api/v1/job/` and `api/v1/project/{project_id}/workflow/{workflow_id}/job/` (\#969, \#1003).
         * Add new `GET` endpoints `api/v1/dataset/` and `api/v1/workflow/` (\#988, \#1003).
         * Add new `GET` endpoint `api/v1/project/{project_id}/dataset/` (\#993).
@@ -20,6 +20,7 @@
         * Move all routes definitions into `fractal_server/app/routes` (\#976).
         * Fix construction of `ApplyWorkflow.workflow_dump`, within apply endpoint (\#968).
         * Fix construction of `ApplyWorkflow` attributes `input_dataset_dump` and `output_dataset_dump`, within apply endpoint (\#990).
+        * Remove `asyncio.gather`, in view of SQLAlchemy2 update (\#1004).
 * Database:
     * Make foreign-keys of `ApplyWorkflow` (`project_id`, `workflow_id`, `input_dataset_id`, `output_dataset_id`) optional (\#927).
     * Add columns `input_dataset_dump`, `output_dataset_dump` and `user_email` to `ApplyWorkflow` (\#927).
@@ -28,10 +29,17 @@
     * Remove `"cascade": "all, delete-orphan"` from `Project.job_list` (\#927).
     * Add `Workflow.job_list` relation (\#927).
     * Do not use `Enum`s as column types (e.g. for `ApplyWorkflow.status`), but only for (de-)serialization (\#974).
+    * Add script for updating DB from 1.4.0 to 1.4.1 (\#1010)
+* App:
+    * Fix missing try/except in sync session (\#1020).
+    * Skip creation of first-superuser when one superuser already exists (\#1006).
 * Runner:
     * Refresh DB objects within `submit_workflow` (\#927).
 * Testing:
     * Improve `test_full_workflow.py` (\#971).
+    * Update `pytest-asyncio` to v0.21 (\#1008).
+    * Fix CI issue related to event loop and asyncpg (\#1012).
+    * Add GitHub Action testing database migrations (\#1010).
```

## `git diff 1.4.0a7...1.4.0a8 CHANGELOG.md`
```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 6f31b5be..28fd9f69 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,13 @@
         * Add new `GET` endpoints `api/v1/job/` and `api/v1/project/{project_id}/workflow/{workflow_id}/job/` (\#969, \#1003).
         * Add new `GET` endpoints `api/v1/dataset/` and `api/v1/workflow/` (\#988, \#1003).
         * Add new `GET` endpoint `api/v1/project/{project_id}/dataset/` (\#993).
+        * Add superuser-only `PATCH /api/v1/job/{job_id}/` endpoint (\#1030).
+        * Move `GET /auth/whoami/` to `GET /auth/current-user/` (\#1013).
+        * Move `PATCH /auth/users/me/` to `PATCH /auth/current-user/` (\#1013, \#1035).
         * Remove `DELETE /auth/users/{id}/` endpoint (\#994).
+        * Remove `GET /auth/users/me/` (\#1013).
+        * Remove `POST` `/auth/forgot-password/`, `/auth/reset-password/`, `/auth/request-verify-token/`, `/auth/verify/` (\#1033).
+        * Move `GET /auth/userlist/` to `GET /auth/users/` (\#1033).
     * New behaviors or responses of existing endpoints:
         * Change response of `/api/v1/project/{project_id}/job/{job_id}/stop/` endpoint to 204 no-content (\#967).
         * Remove `dataset_list` attribute from `ProjectRead`, which affects all `GET` endpoints that return some project (\#993).
@@ -29,10 +35,13 @@
     * Remove `"cascade": "all, delete-orphan"` from `Project.job_list` (\#927).
     * Add `Workflow.job_list` relation (\#927).
     * Do not use `Enum`s as column types (e.g. for `ApplyWorkflow.status`), but only for (de-)serialization (\#974).
+    * Set `pool_pre_ping` option to `True`, for asyncpg driver (\#1037).
     * Add script for updating DB from 1.4.0 to 1.4.1 (\#1010)
 * App:
     * Fix missing try/except in sync session (\#1020).
     * Skip creation of first-superuser when one superuser already exists (\#1006).
+* Dependencies:
+    * Pin sqlalchemy to 1.4.50 and upgrade asyncpg to 0.29.0 (\#1036).
 * Runner:
     * Refresh DB objects within `submit_workflow` (\#927).
 * Testing:
@@ -40,6 +49,8 @@
     * Update `pytest-asyncio` to v0.21 (\#1008).
     * Fix CI issue related to event loop and asyncpg (\#1012).
     * Add GitHub Action testing database migrations (\#1010).
+* Documentation:
+    * Add OAuth2 example endpoints to Web API page (\#1034, \#1038).
```

**Comment by tcompa** [comment]
Notable breaking changes with respect to 1.3.x:
* Move `GET /auth/whoami/` to `GET /auth/current-user/` (\#1013).
* Move `GET /auth/userlist/` to `GET /auth/users/` (\#1033).
* Skip creation of first-superuser when one superuser already exists (\#1006).
